lib/coll.gi

```diff
@@ -1675,497 +1675,233 @@ InstallMethod( ForAnyOp,
-        fi;
-    od;
-    Add( result, gens[i+1]( val1 ) );
-end );
-
-BIND_GLOBAL( "ListXHelp0", function ( result, gens, i )
-    local   gen, val1;
-    while i+1 < Length(gens)  do
-        gen := gens[i+1];
-        if IsFunction( gen )  then
-            gen := gen();
-        fi;
-        if gen = true  then
-            i := i + 1;
-        elif gen = false  then
-            return;
-        elif IsListOrCollection( gen )  then
-            for val1  in gen  do
-                ListXHelp1( result, gens, i+1, val1 );
-            od;
-            return;
-        else
-            Error( "gens[",i+1,"] must be a collection, a list, a boolean, ",
-                   "or a function" );
-        fi;
-    od;
-    Add( result, gens[i+1]() );
-end );
-
 InstallGlobalFunction( ListX, function ( arg )
-    local   result;
-    result := [];
-    ListXHelp0( result, arg, 0 );
-    return result;
+    local f;
+    f := Remove(arg);
+    return FoldLeftX(arg,
+            function(acc, x)
+                Add(acc, CallFuncList(f, x));
+                return acc;
+            end, []);
 end );
 
 
 #############################################################################
 ##
-#M  SetX(<obj>,...)
+#F  FoldLeft( <coll>, <func> )
+#F  FoldLeft( <coll>, <func>, <init> )
 ##
-SetXHelp := function ( result, gens, i, vals, l )
-    local   gen, val;
-    while i+1 < Length(gens)  do
-        gen := gens[i+1];
-        if IsFunction( gen )  then
-            gen := CallFuncList( gen, vals );
-        fi;
-        if gen = true  then
-            i := i + 1;
-        elif gen = false  then
-            return;
-        elif IsListOrCollection( gen )  then
-            for val  in gen  do
-                vals[l+1] := val;
-                SetXHelp( result, gens, i+1, vals, l+1 );
-            od;
-            Unbind( vals[l+1] );
-            return;
-        else
-            Error( "gens[",i+1,"] must be a collection, a list, a boolean, ",
-                   "or a function" );
-        fi;
-    od;
-    AddSet( result, CallFuncList( gens[i+1], vals ) );
-end;
-MAKE_READ_ONLY_GLOBAL( "SetXHelp" );
-
-BIND_GLOBAL( "SetXHelp2", function ( result, gens, i, val1, val2 )
-    local   gen, vals, val3;
-    while i+1 < Length(gens)  do
-        gen := gens[i+1];
-        if IsFunction( gen )  then
-            gen := gen( val1, val2 );
-        fi;
-        if gen = true  then
-            i := i + 1;
-        elif gen = false  then
-            return;
-        elif IsListOrCollection( gen )  then
-            vals := [ val1, val2 ];
-            for val3  in gen  do
-                vals[3] := val3;
-                SetXHelp( result, gens, i+1, vals, 3 );
-            od;
-            Unbind( vals[3] );
-            return;
+InstallGlobalFunction( FoldLeft,
+    function( arg )
+    local tnum, C, func, result, i, l;
+    l := Length( arg );
+    if l < 2 or l > 3 or not IsFunction(arg[2]) then
+      Error( "usage: FoldLeft( <C>, <func>[, <init>] )" );
```

Comment:
TODO: should `FoldLeft` also get an abort mechanism, like `FoldLeftX`? How should it work? The one for `FoldLeftX` is a bit adhoc: it works great for the few applications we have, but it's not very general...

What do other languages do for this, if they do anything at all?